spec/mem/tree.md

```diff
@@ -0,0 +1,564 @@
+                Permission::Reserved => Permission::Reserved,
+                Permission::Active => Permission::Frozen,
+                Permission::Frozen => Permission::Frozen,
+                Permission::Disabled => Permission::Disabled,
+            }
+        )
+    }
+
+     fn foreign_write(permission: Permission) -> Result<Permission> {
+        ret(
+            match permission {
+                Permission::Reserved => Permission::Disabled,
+                Permission::Active => Permission::Disabled,
+                Permission::Frozen => Permission::Disabled,
+                Permission::Disabled => Permission::Disabled,
+            }
+        )
+    }
+
+    fn permission_transition(curr: Permission, access_pattern: AccessKind, node_relation: NodeRelation) -> Result<Permission> {
+        match (access_pattern, node_relation) {
+            (AccessKind::Read, NodeRelation::Child) => Self::child_read(curr),
+            (AccessKind::Write, NodeRelation::Child) => Self::child_write(curr),
+            (AccessKind::Read, NodeRelation::Foreign) => Self::foreign_read(curr),
+            (AccessKind::Write, NodeRelation::Foreign) => Self::foreign_write(curr),
+        }
+    }
+}
+```
+
+We define the tree structure to track reborrows in each allocation. The tree consists of several nodes, each corresponding to a borrow tag (i.e., a reborrow). Each node has a parent and a list of children. Additionally, we track each tag's permissions on memory locations (bytes) within the allocation.
+
+```rust
+pub struct Tree {
+    root_tag: BorTag, 
+    /// Map a borrow tag to a node
+    nodes: Map<BorTag, Node>,
+}
+
+pub struct Node {
+    /// Borrow tag of the parent node
+    parent: Option<BorTag>, 
+    /// Borrow tags of the children node
+    children: List<BorTag>,
+    /// Permission for each location
+    permissions: Map<Address, Permission>,
+}
+
+pub enum NodeRelation {
+    Child, 
+    Foreign,
+}
+
+impl Tree {
+    /// Insert a node into the tree
+    fn insert_node(&mut self, tag: BorTag, node: Node) {
+        self.nodes.insert(tag, node);
+    }
+}
+```
+
+To update the tree during each memory access, each node in the tree can be divided into two disjoint sets: Child and Foreign. The Child set includes the node itself and all its descendants, while the Foreign set contains all other nodes. When a node is accessed, we must update and check permissions for each node in the tree based on its relative position to the accessed node.
+
+```rust
+impl TreeBorrowsAllocation {
+    /// Recusively do permission transition on the `curr` node and all its descendants.
+    /// `base` means the node that is actually memory accessed.
+    /// This method will throw UBs, representing the undefined behavior captured by Tree Borrows.
+    fn node_access(
+        &mut self, 
+        curr: BorTag,
+        access_pattern: AccessKind,
+        ptr: Pointer<TreeBorrowsProvenance>,
+        size: Size,
+        base: BorTag,
```

Comment:
Maybe call this "access_tag" or so, `base` is not very self-explaining.

"curr" would then become "curr_tag".